#Summary of WordNet 

WordNet is a word and lexial database that consist of words and all of its forms. WordNet also defines and gives out exmaples of how to use these words. 

In [6]:
import nltk
nltk.download('wordnet')
nltk.download('omw-1.4')
from nltk.corpus import wordnet as wn
#Select a noun. Output all synsets.
wn.synsets('work')

[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...


[Synset('work.n.01'),
 Synset('work.n.02'),
 Synset('employment.n.02'),
 Synset('study.n.02'),
 Synset('work.n.05'),
 Synset('workplace.n.01'),
 Synset('oeuvre.n.01'),
 Synset('work.v.01'),
 Synset('work.v.02'),
 Synset('work.v.03'),
 Synset('function.v.01'),
 Synset('work.v.05'),
 Synset('exercise.v.03'),
 Synset('make.v.36'),
 Synset('work.v.08'),
 Synset('work.v.09'),
 Synset('work.v.10'),
 Synset('bring.v.03'),
 Synset('work.v.12'),
 Synset('cultivate.v.02'),
 Synset('work.v.14'),
 Synset('influence.v.01'),
 Synset('work.v.16'),
 Synset('work.v.17'),
 Synset('work.v.18'),
 Synset('work.v.19'),
 Synset('shape.v.02'),
 Synset('work.v.21'),
 Synset('knead.v.01'),
 Synset('exploit.v.01'),
 Synset('solve.v.01'),
 Synset('ferment.v.03'),
 Synset('sour.v.01'),
 Synset('work.v.27')]

In [ ]:
#Select one synset from the list of synsets. Extract its definition, usage examples, and lemmas.
wn.synset('work.n.01').definition()


'activity directed toward making or doing something'

In [ ]:
wn.synset('work.n.01').examples()


['she checked several points needing further work']

In [ ]:

wn.synset('work.n.01').lemmas()


[Lemma('work.n.01.work')]

In [16]:

work = wn.synset('work.n.01')
hyper = work.hypernyms()[0]
t = wn.synset('entity.n.01')
while hyper:
    print(hyper)
    if hyper == t:
        break
    if hyper.hypernyms():
        hyper = hyper.hypernyms()[0]

Synset('activity.n.01')
Synset('act.n.02')
Synset('event.n.01')
Synset('psychological_feature.n.01')
Synset('abstraction.n.06')
Synset('entity.n.01')


#Observation

I feel that WordNet works better with nouns, as sometimes it has multiple definitions of a single word. Unlike other word forms, where it might has very few definitions to none. Also, I noticed that WordNet works very well for noun words, as it provides a hierarchy of the word (until the word entity). Also, WordNet can define a word, state the lemmatizations of the word, and provide examples of the word.  

The noun I used was work, and the definition I got is "activity directed toward making or doing something," which I believe to be accurate. The examples I got is "she checked several points needing further work," which I also believe is accurate. And the lemma I received is 'work.n.01.work'

For the WordNet hierarchy, I received these words: activity, act, event, psychological_feature, abstraction, and entity, which are also similar to dance. 

Write a couple of sentences observing the way that WordNet is organized for nouns.

In [ ]:
#Output the following (or an empty list if none exist): hypernyms, hyponyms, meronyms, holonyms, antonym.
wn.synset('work.n.01').hypernyms()

[Synset('activity.n.01')]

In [ ]:
wn.synset('work.n.01').hyponyms()

[Synset('action.n.10'),
 Synset('busywork.n.01'),
 Synset('care.n.01'),
 Synset('coursework.n.01'),
 Synset('duty.n.02'),
 Synset('heavy_lifting.n.01'),
 Synset('housewifery.n.01'),
 Synset('housework.n.01'),
 Synset('investigation.n.02'),
 Synset('ironing.n.02'),
 Synset('job.n.06'),
 Synset('job.n.07'),
 Synset('labor.n.02'),
 Synset('logging.n.01'),
 Synset('loose_end.n.01'),
 Synset('mission.n.04'),
 Synset('nightwork.n.01'),
 Synset('operation.n.07'),
 Synset('paperwork.n.01'),
 Synset('service.n.01'),
 Synset('shining.n.01'),
 Synset('spadework.n.01'),
 Synset('subbing.n.01'),
 Synset('timework.n.01'),
 Synset('undertaking.n.01'),
 Synset('wash.n.02'),
 Synset('welfare_work.n.01')]

In [ ]:
wn.synset('work.n.01').part_meronyms()

[]

In [ ]:
wn.synset('work.n.01').part_holonyms()

[]

In [ ]:
w = wn.synset('work.n.01')
w.lemmas()[0].antonyms()

[]

In [ ]:
#Select a verb. Output all synsets.
wn.synsets('dance')

[Synset('dance.n.01'),
 Synset('dance.n.02'),
 Synset('dancing.n.01'),
 Synset('dance.n.04'),
 Synset('dance.v.01'),
 Synset('dance.v.02'),
 Synset('dance.v.03')]

In [ ]:
wn.synset('dance.v.01').definition()

'move in a graceful and rhythmical way'

In [ ]:
wn.synset('dance.v.01').examples()

['The young girl danced into the room']

In [ ]:
wn.synset('dance.v.01').lemmas()

[Lemma('dance.v.01.dance')]

In [ ]:
# traversing 
act = wn.synset('dance.v.01')
hyper = act.hypernyms()[0]


for i in range(10):
    print(hyper)
    if hyper.hypernyms():
        hyper = hyper.hypernyms()[0]

Synset('move.v.03')
Synset('move.v.03')
Synset('move.v.03')
Synset('move.v.03')
Synset('move.v.03')
Synset('move.v.03')
Synset('move.v.03')
Synset('move.v.03')
Synset('move.v.03')
Synset('move.v.03')



#Observation

I feel that WordNet does not work well for verb as the way it dose for nouns. Even though a word might have multiple definitions of a verb, the definitions are not as diverse as the noun definitions. However, WordNet provides more options for verbs than other word forms that are not nouns. Also, WordNet provides a hierarchy of the verb, however, WordNet cannot traverse a verb until entity. But WordNet can provide other words with the same meaning as the verb (or similar). Also, WordNet can define a verb, state the lemmatizations of its, and provide some examples.  

The verb I used was dance, and the definition I got is "move in a graceful and rhythmical way," which I believe to be accurate. The examples I got is "The young girl danced into the room," which I also believe is accurate. And the lemma I received is 'dance.v.01.dance'

For the WordNet hierarchy, I received this word 'move.v.03', which also similar to dance. 

In [ ]:
#  morphy 
wn.morphy('acting', wn.ADJ)

'act'

In [ ]:
#  morphy 
wn.morphy('acting', wn.VERB)

'act'

In [ ]:
#  morphy 
wn.morphy('acting', wn.NOUN)

'acting'

In [ ]:
#  morphy 
wn.morphy('acting', wn.ADJ_SAT)

'acting'

In [ ]:
#  morphy 
wn.morphy('acting', wn.ADV)

In [4]:
#Select two words and find similarities 
wn.synsets('work')



[Synset('work.n.01'),
 Synset('work.n.02'),
 Synset('employment.n.02'),
 Synset('study.n.02'),
 Synset('work.n.05'),
 Synset('workplace.n.01'),
 Synset('oeuvre.n.01'),
 Synset('work.v.01'),
 Synset('work.v.02'),
 Synset('work.v.03'),
 Synset('function.v.01'),
 Synset('work.v.05'),
 Synset('exercise.v.03'),
 Synset('make.v.36'),
 Synset('work.v.08'),
 Synset('work.v.09'),
 Synset('work.v.10'),
 Synset('bring.v.03'),
 Synset('work.v.12'),
 Synset('cultivate.v.02'),
 Synset('work.v.14'),
 Synset('influence.v.01'),
 Synset('work.v.16'),
 Synset('work.v.17'),
 Synset('work.v.18'),
 Synset('work.v.19'),
 Synset('shape.v.02'),
 Synset('work.v.21'),
 Synset('knead.v.01'),
 Synset('exploit.v.01'),
 Synset('solve.v.01'),
 Synset('ferment.v.03'),
 Synset('sour.v.01'),
 Synset('work.v.27')]

In [5]:
wn.synsets('job')

[Synset('occupation.n.01'),
 Synset('job.n.02'),
 Synset('job.n.03'),
 Synset('job.n.04'),
 Synset('job.n.05'),
 Synset('job.n.06'),
 Synset('job.n.07'),
 Synset('problem.n.01'),
 Synset('job.n.09'),
 Synset('job.n.10'),
 Synset('job.n.11'),
 Synset('job.n.12'),
 Synset('caper.n.03'),
 Synset('job.v.01'),
 Synset('subcontract.v.01'),
 Synset('job.v.03'),
 Synset('speculate.v.04')]

In [6]:
w1 = wn.synset('workplace.n.01')
w2 = wn.synset('problem.n.01')

#wu-palmer:
wn.wup_similarity(w1, w2)


0.14285714285714285

In [9]:
from nltk.wsd import lesk
print(lesk('workplace', 'problem'))

Synset('trouble.n.01')



#Observations

First, I found the synsets of two words, and choose one synset of each one. Then I saved the synsets in variables w1 and w2. After that I tested the words similarities using Wu-Palmer and Lesk functions. 
I have noticed that Wu-Palmer function provide a fraction number between 0 and 1, and the scale is from 0 to 1, 0 means that the words are not similar, and 1 means that the words are very similar. The Wu-Palmer score I got was 0.142, which means that the words I picked were not very similar.

The Lesk functions provides word/s that are similar to the picked words. The words I used for both Wu-Palmer and Lesk functions are: workplace and problem. The word I got from the Lesk function is trouble. 



#SentiWordNet Summary

SentiWordNet is considered a library of WordNet, that provides three results fora word (synset) about positive, negative, and objective indication of the word. For example, we can use words that has negative meaning such as 'mad' and SentiWordNet would provide us with three score about how negative, positive and objective the word is, and the negative result would ultimately be higher than the rest. However, the word 'happy' would provide us with a higher positive score than negative. 

In [12]:
wn.synsets('thrilling')

[Synset('thrill.v.01'),
 Synset('thrill.v.02'),
 Synset('shudder.v.02'),
 Synset('exhilarate.v.01'),
 Synset('electrifying.s.01'),
 Synset('thrilling.s.02')]

In [1]:
# Select an emotionally charged word.
#Find its senti-synsets and output the polarity scores for each word. 
#Make up a sentence. Output the polarity for each word in the sentence. 
import nltk
nltk.download('sentiwordnet')
from nltk.corpus import sentiwordnet as swn



[nltk_data] Downloading package sentiwordnet to /root/nltk_data...
[nltk_data]   Unzipping corpora/sentiwordnet.zip.


In [17]:

import nltk
nltk.download('sentiwordnet')
nltk.download('wordnet')
from nltk.corpus import sentiwordnet as swn
sentl = list(swn.senti_synsets('electrifying')) 
for i in sentl:
    print(i)

    print("objective: ", i.obj_score())

<electrify.v.01: PosScore=0.25 NegScore=0.125>
objective:  0.625
<electrify.v.02: PosScore=0.0 NegScore=0.0>
objective:  1.0
<electrify.v.03: PosScore=0.0 NegScore=0.0>
objective:  1.0
<electrifying.s.01: PosScore=0.25 NegScore=0.5>
objective:  0.25


[nltk_data] Downloading package sentiwordnet to /root/nltk_data...
[nltk_data]   Package sentiwordnet is already up-to-date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


In [21]:
#Make up a sentence. Output the polarity for each word in the sentence. 
s = 'I love the weather today'
negative = 0
positive  = 0
t = s.split()
for token in t:
    synl = list(swn.senti_synsets(token))
    if synl:
        word = synl[0]
        negative += word.neg_score()
        positive += word.pos_score()
    
print("negative scores: ", negative)
print('positive scores: ', positive)


negative scores:  0.0
positive scores:  0.75




#Observations

I selected the word 'electrifying' and created a list of senti_synsets to loop through. Looping through this the senti_synsets provided me with the word and the polarity score for each one. The results I got indicated that the word has more an objective meaning, positive, then negative. 
For the sentence, I used this "I love the weather today" and also used a list of senti_synsets of tokens. The results I reced showed me that the sentence is positive and not negative.

#Collocation
 
collocation means the likelihood of two exacte words (no other words with similar meaning) occuring together. And NLKT has nine collocation options named as text1 through 9. Collocation can be created or we can use the collocation provided to us from NLKT library. Then we can use the data from the collocation to calculate mutual information. 

In [11]:

#Output collocations for text4, the Inaugural corpus. Select one of the collocations identified by NLTK. 
#Calculate mutual information. 

import nltk
nltk.download('inaugural')
nltk.download('nps_chat')
nltk.download('webtext')
nltk.download('treebank')
nltk.download('stopwords')
nltk.download('gutenberg')
nltk.download('genesis')
  
from nltk.book import *
text4.collocations()


[nltk_data] Downloading package inaugural to /root/nltk_data...
[nltk_data]   Package inaugural is already up-to-date!
[nltk_data] Downloading package nps_chat to /root/nltk_data...
[nltk_data]   Package nps_chat is already up-to-date!
[nltk_data] Downloading package webtext to /root/nltk_data...
[nltk_data]   Package webtext is already up-to-date!
[nltk_data] Downloading package treebank to /root/nltk_data...
[nltk_data]   Package treebank is already up-to-date!
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!
[nltk_data] Downloading package gutenberg to /root/nltk_data...
[nltk_data]   Package gutenberg is already up-to-date!
[nltk_data] Downloading package genesis to /root/nltk_data...
[nltk_data]   Unzipping corpora/genesis.zip.


*** Introductory Examples for the NLTK Book ***
Loading text1, ..., text9 and sent1, ..., sent9
Type the name of the text or sentence to view it.
Type: 'texts()' or 'sents()' to list the materials.
text1: Moby Dick by Herman Melville 1851
text2: Sense and Sensibility by Jane Austen 1811
text3: The Book of Genesis
text4: Inaugural Address Corpus
text5: Chat Corpus
text6: Monty Python and the Holy Grail
text7: Wall Street Journal
text8: Personals Corpus
text9: The Man Who Was Thursday by G . K . Chesterton 1908
United States; fellow citizens; years ago; four years; Federal
Government; General Government; American people; Vice President; God
bless; Chief Justice; one another; fellow Americans; Old World;
Almighty God; Fellow citizens; Chief Magistrate; every citizen; Indian
tribes; public debt; foreign nations


In [22]:
t = ' '.join(text4.tokens)
t[:67]

'Fellow - Citizens of the Senate and of the House of Representatives'

In [24]:
import math
voc = len(set(text4))
hg = t.count('and to')/voc
print("p(and to) = ",hg )
h = t.count('and')/voc
print("p(and) = ", h)
g = t.count('to')/voc
print('p(to) = ', g)
pmi = math.log2(hg / (h * g))
print('pmi = ', pmi)

p(and to) =  0.017955112219451373
p(and) =  0.5833416458852868
p(to) =  0.565286783042394
pmi =  -4.198929370502643




#Observation 
I used this sentence 'and to' to calculate the mutual information and receive the PMI results. the occurrence of the sentence was 0.0179, which is very small, but the occurrence of individual words 'and' = 0.5833 and 'to' = 0.565, which is very high, however, the likelihood of them used together is very small. 